RIPS/rip-9995.md

```diff
@@ -0,0 +1,199 @@
+often lead to unnecessary technical debt and are worth avoiding if possible.
+
+Additionally, method selectors provide very weak indication of the contract developer's decision to become a
+participant in Native Account Abstraction transaction flow.
+The chance of accidentally exposing a function with a colliding 4 byte
+method identifier that returns a valid 4 byte magic to indicate approval is pretty low, but a malicious developer
+can easily hide such a function giving it an innocent name, making it hard to spot a Native Account Abstraction entity.
+
+This issue to some extent is also present in ERC-4337.
+
+As an alternative, if Native Account Abstraction is to be implemented in coordination with EIP-3540,
+relying on the concept of "code sections" it introduces is a superior approach.
+
+## Specification
+
+### Validation and PostTransaction code sections
+
+The EOF format describes a separation of 'code' and 'data' sections,
+marked with `kind_code (0x02)` and `kind_data (0x04)` kind markers.
+
+The contract that may have a role in an Account Abstraction transaction, either as a Sender, a Paymaster or a Deployer,
+has to contain a section marked with the following kind marker:
+
+```
+kind_abstraction = 0x05
+```
+
+This section is equivalent to a code section.
+
+Its code can be executed during a regular transaction execution and has no special effects.
+If it is the first code section of a contract, it can act as an entry point during regular transaction execution.
+
+Only a single section of `kind_abstraction` is allowed in a contract.
+This rule is validated during contract creation.
+
+### Execution entry point for Account Abstraction transaction type participant entity (Sender, Paymaster and Deployer)
+
+During a regular contract code execution, its behaviour is defined as follows by EIP-3540:
+```
+Execution starts at the first byte of code section 0, and pc is set to 0
+```
+
+However, if a contract is referenced in an `AA_TX_TYPE` transaction as a Sender, Paymaster or a Deployer,
+execution starts at the first byte of code section with `kind_abstraction` marker, and `pc` is set to 0.
+
+If the specified contract does not contain such a section, or is not an EOF contract, the transaction is not valid.
+
+### Encoding inputs for different execution frames
+
+We define the following "action" identifiers:
+
+```
+sender_deployment = 0x01
+sender_validation = 0x02
+paymaster_validation = 0x03
+paymaster_posttx = 0x04
+```
+
+#### Sender Deployment
+
+Inputs to the `Deployer` contract are not defined by the protocol and are controlled by the `deployerData` parameter.
+
+The sender deployment frame MUST result in the `sender` address becoming initialized with contract code.
+
+#### Sender Validation
+
+Inputs to the `Sender` validation section are defined by the protocol as a packed encoding of the following data:
+```
+uint256 version, uint256 action, bytes32 txHash, bytes transaction
```

Comment:
There has been a number of changes is the transaction structure in Ethereum in the last couple of years. Do you think we can abandon the `version` field and just request new section IDs in case we need to modify the transaction structure?